### ISSUE TITLE
Artificial classDef confuses getCanonicalTypeForIdentifier()

### ISSUE THREAD

**Issue by marcalff** [doxygen/doxygen#8050]
See attached .tar.gz for full code, with reproducible script.
See file reproduce_bug.ksh
See file analysis.txt

///

PROBLEM:
========

The following code:

```
#include <vector>

static bool fetch_keys(int *, std::vector<int> *);

static bool fetch_keys(int *keyring,
                       std::vector<int> *keyring_keys) {
  return false;
}
```

causes the following error

```
/home/malff/EXTERNAL/doxygen/bug_1/storage/perfschema/table_keyring_keys.cc:38: warning: no matching file member found for
static bool fetch_keys(int *keyring, std::vector< int > *keyring_keys)
Possible candidates:
 'static bool fetch_keys(int *, std::vector< int > *)' at line 38 of file /home/malff/EXTERNAL/doxygen/bug_1/storage/perfschema/table_keyring_keys.cc
```

but only if doxygen process -- other -- input files also.

This can be reproduced using Doxyfile 'Doxyfile-both'.

When using Doxyfile 'Doxyfile-single', which parses exactly
the same function fetch_keys() in file
storage/perfschema/table_keyring_keys.cc,
no error is reported.

The expected result is no error, in all cases.

ROOT CAUSE:
===========

When generating documentation for (user input) fetch_keys() ...

1. Function findGlobalMember() in doxygen.cpp reaches "4. Try to add member 'fetch_keys' to scope '' and then calls matchArguments2(), which fails to match the input.
2. Function matchArguments2() in util.cpp loops for each arguments, and calls a nested matchArgument2() function.
    For the second parameter (`std::vector< int > *`), matchArgument2() fails.
3. Function matchArgument2() in util.cpp calls extractCanonicalArgType() for both source and destination, and compare the results.

   For one call, extractCanonicalArgType() returns
     "std::vector< int >"
   as expected.

   For the other call, extractCanonicalArgType returns
     "vector< int >"
   instead, causing the match to fail.
4. For the failing call, extractCanonicalArgType() calls getCanonicalTypeForIdentifier(), which returns a wrong result.
5. For the failing call, getCanonicalTypeForIdentifier() calls getResolvedClass() to get a class definition, and then build a result using
   cd->qualifiedName() + templSpec

   The problem is that the
     const ClassDef *cd;
   object found has the following attributes:
   - cd->name() is "std::vector", ok
   - cd->qualifiedName() is "vector" instead of "std::vector", causing the bug.

   This class definition happens to have the interesting property:
     - cd->isArtificial() is true
     - cd->getDefFileName() point to the code where this class was seen.

   For the user input,
     cd->getDefFileName() pointed to file sql/item_geofunc_internal.cc
   instead of the expected C++ header files (`#include <vector>`).

   Because of this:
   - `std::vector<int>` is normalized as `vector<int>`,
   - the type comparison of arguments of fecth_keys() fails,
   - dogygen reports an error, even when the source code is correct.
  
   Now, removing the input file sql/item_geofunc_internal.cc from Doxyfile INPUT (see reproduce_bug.ksh) clears the issue,
and file storage/perfschema/table_keyring_keys.cc is generated properly.
6. The class definition added in the class definition dictionary, for the artificial std::vector ClassDef, comes from parsing this code,
in INPUT file sql/item_geofunc_internal.cc

   ```
   #include <vector>

   template <typename T>
   class Pointer_vector : public std::vector<T *> {
     typedef std::vector<T *> parent;

    public:
     ~Pointer_vector() {
          for (typename parent::iterator i = this->begin(); i != this->end(); ++i)
           delete (*i);
        }
   };
   ```

   Somehow, adding artificial classes for std::vector confuses doxygen when computing canonical types in getCanonicalTypeForIdentifier().

OTHER
=====

This affects the doxygen build of MySQL.

Feel free to contact me if further investigations are needed.

email: marc.alff@oracle.com

Regards,
-- Marc Alff

[bug_artificial.tar.gz](https://github.com/doxygen/doxygen/files/5262275/bug_artificial.tar.gz)

**Comment by albert-github** [comment]
Some investigations gave that the version:
```
Commit: cd581388f3d013c501e3cefbaf3e81cf93d46fcb [cd58138]
Committer: Dimitri van Heesch
issue #7706: Md5 hash does not match for two different runs (part 2)
```
was still OK, but the next commit:
```
Commit: 9e2b6a2dac870b5daabde118c5327281e3f79ba2 [9e2b6a2]
Commit Date: Friday, April 24, 2020 9:03:09 AM
Performance improvements after profiling

In some cases performance dropped when upgrading from version 1.8.16 to 1.8.17 or 1.8.18.
With these changes the performance should be back to normal again.
```
shows the problem.

**Comment by doxygen** [comment]
@marcalff I think the problem is that doxygen does not know what `std::vector` is (i.e. it is not part of the input). 

An easy solution it to set `BUILTIN_STL_SUPPORT` to `YES`, so doxygen internally adds definitions for the most common STL classes. Let me if that solves the problem for you.

**Comment by marcalff** [comment]
To @doxygen,

Hi Dimitri, Albert,

Thanks for the suggestion.

Please see blow, for:
- a summary of known facts, to understand the problem,
- some analysis and comments, for discussion.

Regards,
-- Marc Alff

PART I
======

Based on more testing, this is the current picture:

1) The code base used (MySQL) contains sometimes complex code, involving STL,
for example class std:vector<T>.

2) The STL headers are not part of the INPUT file,
so that Doxygen does not know about its classes,
as you pointed out.

Adding the STL headers in the INPUT is not an option,
as we want to document only our code base,
not polluted with extra symbols from third parties.

3) With Doxygen 1.8.18, the build is clean (*).

This is confirmed both:
- with BUILTIN_STL_SUPPORT=NO
- with BUILTIN_STL_SUPPORT=YES

The error log is identical.

(*) Here, 'clean' means issue #8050 does not cause spurious errors,
there are still unrelated things to fix in our code/doc.

4) With Doxygen 1.8.19 and BUILTIN_STL_SUPPORT=NO,
the Doxygen reports a lot of spurious errors,
all related to non matching prototypes that uses STL code.

The initial report pointed at a specific case involving "std::vector",
with reproduction steps, but the number of errors is much bigger,
and involves many different classes.

With Doxygen 1.8.19 and BUILTIN_STL_SUPPORT=YES,
these spurious errors are fixed.

5) With Doxygen 1.8.20, same results as 1.8.19:
- the build with BUILTIN_STL_SUPPORT=NO has spurious errors
- the build with BUILTIN_STL_SUPPORT=YES is clean.

So far, one could conclude that BUILTIN_STL_SUPPORT=YES is the proper
solution, but I think there is more to it : now on to part II.

PART II
=======

If I understand correctly, Doxygen finds knowledge about STL classes,
which are not part of the INPUT, by either:
- current usage in the code
- explicit seed with BUILTIN_STL_SUPPORT=YES.

A) Creating artificial classes from code usage:

As reported initially, the following code seen in INPUT:

<pre>
#include <vector>

template <typename T>
class Pointer_vector : public std::vector<T *> {
  typedef std::vector<T *> parent;

 public:
  ~Pointer_vector() {
    for (typename parent::iterator i = this->begin(); i != this->end(); ++i)
      delete (*i);
  }
};
</pre>

causes Doxygen to create an "Artificial" class for "std::vector".

In case of Doxygen 1.8.20,
debugging showed that the
  const ClassDef *cd;
object created has the following attributes:
- cd->name() is "std::vector", ok
- cd->qualifiedName() is "vector" instead of "std::vector", causing the bug.

Once this artificial class is added to the dictionary,
the dictionary is poisoned with incorrect data,
causing failures when parsing unrelated INPUT files that uses the same
class.

B) Using BUILTIN_STL_SUPPORT=YES causes artificial classes to be added to seed
the dictionary for STL.

This is my understanding from reading doxygen/src/stlsupport.cpp.

So, Doing B) seeds correct data for "std::vector",
avoiding the need to to this using path A).

As a result, BUILTIN_STL_SUPPORT=YES effectively hides the bug in A),
and the build is clean again.

BUILTIN_STL_SUPPORT=YES is a work around, not a fix.

Some additional comments below:

C) BUILTIN_STL_SUPPORT=YES only works for, well, STL.

If the original code involving "std::vector" is renamed to use "foo::bar"
instead, I expect the same bug in A) to re appear,
but without a work around this time..

D) BUILTIN_STL_SUPPORT=YES is a very loose approximation of the STL.

Reading the code in doxygen/src/stlsupport.cpp,
it seems definitions from different versions of C++ (C++11, C++17, C++20, etc)
are all added to the dictionary, even when the project does not even uses
the corresponding version of C++.

In the long term, there is the risk that a build for a project using say C++11
could be confused by artificial definitions from say C++20 added by Doxygen
itself.

In the long term also, this scheme does not cover removal (in C++2048 ?) of classes
deprecated a long time ago.

Given that Doxygen does not currently know which flavor of STL is used,
the configuration variable BUILTIN_STL_SUPPORT possibly needs
to change from a basic 'YES' / 'NO' flag to something like:
'NO' / 'C++11' / 'C++17' / 'C++20' / 'YES' (meaning all flavors ?) ...

BUILTIN_STL_SUPPORT=YES is probably good enough for now,
just commenting on this as it popped up while reading the code.

E) BUILTIN_STL_SUPPORT=YES adds code that does not exists in STL.

For example,
still from reading doxygen/src/stlsupport.cpp,
class "std::forward_list" is marked as having iterators,
so it gets:
- "::iterator" and "::const_iterator" (OK)
- "::reverse_iterator" and "::const_reverse_iterator" (for a forward list ?)

The likelihood of this causing an actual error in some INPUT code base is
next to zero, but this shows that seeding artificial class definitions from
hard coded data can have some risks too.

------

In conclusion:

BUILTIN_STL_SUPPORT=YES is a workaround, which I will probably use anyway.
Thanks for suggesting it.

BUILTIN_STL_SUPPORT=YES hides another bug, which still exists, as I
understand.

This bug #8050 is a regression, introduced in 8.0.19 and not fixed since.
8.0.18 was working just fine, even with BUILTIN_STL_SUPPORT=NO.

From previous debugging (which I hope actually helps), the root cause seems related to the ClassDef::qualifiedName() of artificial classes found while parsing the INPUT.

Regards,
-- Marc Alff

**Comment by doxygen** [comment]
@marcalff Thanks for your detailed analysis. I agree with your conclusion. It is in general hard to properly deal with missing information, but in the case illustrated by your example, doxygen should have introduced `std` as an artificial container for `vector` as it also does in other situations. I've added this in the referenced commit. Let me know if it helps.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.0.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
